### ISSUE TITLE
`tag_recv_nb` takes `UCXWorker` or `UCXEndpoint`

### ISSUE THREAD

**Issue by jakirkham** [rapidsai/ucx-py#619]
Currently [`tag_recv_nb`]( https://github.com/rapidsai/ucx-py/blob/5e649abdfb4d339b8c312db382330e11bdefda0a/ucp/_libs/ucx_api.pyx#L853 ) takes [a `UCXWorker`]( https://github.com/rapidsai/ucx-py/blob/5e649abdfb4d339b8c312db382330e11bdefda0a/ucp/_libs/ucx_api.pyx#L854 ) or [a `UCXEndpoint`]( https://github.com/rapidsai/ucx-py/blob/5e649abdfb4d339b8c312db382330e11bdefda0a/ucp/_libs/ucx_api.pyx#L863 ). Though it looks like [we always supply a `UCXEndpoint` when calling `tag_recv_nb`]( https://github.com/rapidsai/ucx-py/blob/c60f43b3cd9c304053e50a0e4d6d06500e450b1d/ucp/comm.py#L78 ). Given this, do we actually have a need to supply both? Or should we simplify the API to only take a `UCXEndpoint`? Mentioning this in part as other send/recv calls do the latter as well.

**Comment by jakirkham** [comment]
Related it appears `_inflight_msgs` is defined in [`UCXWorker`]( https://github.com/rapidsai/ucx-py/blob/5e649abdfb4d339b8c312db382330e11bdefda0a/ucp/_libs/ucx_api.pyx#L309 ) and [`UCXEndpoint`]( https://github.com/rapidsai/ucx-py/blob/5e649abdfb4d339b8c312db382330e11bdefda0a/ucp/_libs/ucx_api.pyx#L453 ). Do both classes need to track this? Or would it be possible to track this in just one of them?

**Comment by jakirkham** [comment]
As [`UCXEndpoint` has an associated `UCXWorker`]( https://github.com/rapidsai/ucx-py/blob/5e649abdfb4d339b8c312db382330e11bdefda0a/ucp/_libs/ucx_api.pyx#L456 ), we might be able to just take a `UCXEndpoint` in `tag_recv_nb` and retrieve the associated `UCXWorker` from it where needed.

**Comment by madsbk** [comment]
This is on purpose, `tag_recv_nb()` use to take a `UCXEndpoint` but as part of the effort of separating the low-level UCX API from the higher level Python/Async API, `tag_recv_nb()` now takes a `UCXWorker`, which match the [UCX API](http://openucx.github.io/ucx/api/latest/html/group___u_c_p___c_o_m_m.html#gaeb24cdfc33dfb2886551b51843aa6304).

**Comment by jakirkham** [comment]
Yeah that makes sense. Was looking at the UCX API and noticed that as well. Just wondering if there's an opportunity here to deduplicate things a bit. WDYT?

**Comment by jakirkham** [comment]
Based on today's meeting, it sounds like @MattBBaker's works will help us solve this by using addresses. Though please correct me if I misunderstood anything, Matt 🙂

**Comment by MattBBaker** [comment]
Just now noticed this, but I don't think so. The use case for UCXWorker address is that it is created, then sent over the network where the otherside gets it as a buffer that is a magic byte string. That byte string is fed into the `ucp_create_ep()` with a flag that says, "Your input data is to be interpreted as a ucp_address_t" and you get an EP from that.
